src/main/java/com/oath/halodb/HaloDBInternal.java

```diff
@@ -457,39 +524,51 @@ private long buildInMemoryIndex(HaloDBOptions options) throws IOException {
-                    // a newer version of the record, replace existing record in index with newer one.
-                    inMemoryIndex.put(key, new InMemoryIndexMetaData(fileId, valueOffset, valueSize, sequenceNumber));
-
-                    // update stale data map for the previous version.
-                    addFileToCompactionQueueIfThresholdCrossed(existing.getFileId(), Utils.getRecordSize(key.length, existing.getValueSize()));
+                InMemoryIndexMetaData metaData = new InMemoryIndexMetaData(fileId, valueOffset, valueSize, sequenceNumber);
+
+                if (!inMemoryIndex.putIfAbsent(key, metaData)) {
+                    while (true) {
+                        InMemoryIndexMetaData existing = inMemoryIndex.get(key);
+                        if (existing.getSequenceNumber() >= sequenceNumber) {
+                            // stale data, update stale data map.
+                            addFileToCompactionQueueIfThresholdCrossed(fileId, recordSize);
+                            break;
+                        }
+                        if (inMemoryIndex.replace(key, existing, metaData)) {
+                            // update stale data map for the previous version.
+                            addFileToCompactionQueueIfThresholdCrossed(existing.getFileId(), Utils.getRecordSize(key.length, existing.getValueSize()));
+                            inserted++;
+                            break;
+                        }
+                    }
+                } else {
                     inserted++;
                 }
-                else {
-                    // stale data, update stale data map.
-                    addFileToCompactionQueueIfThresholdCrossed(fileId, recordSize);
-                }
             }
             logger.debug("Completed scanning index file {}. Found {} records, inserted {} records", fileId, count, inserted);
             indexFile.close();
+
+            return maxSequenceNumber;
         }
+    }
 
-        // Scan all the tombstone files and remove records from index.
-        File[] tombStoneFiles = dbDirectory.listTombstoneFiles();
-        logger.info("About to scan {} tombstone files ...", tombStoneFiles.length);
-        for (File file : tombStoneFiles) {
-            TombstoneFile tombstoneFile = new TombstoneFile(file, options, dbDirectory);
+    class ProcessTombstoneFileTask implements Callable<Long> {
+        private final TombstoneFile tombstoneFile;
+
+        public ProcessTombstoneFileTask(TombstoneFile tombstoneFile) {
+            this.tombstoneFile = tombstoneFile;
+        }
+
+        @Override
+        public Long call() throws IOException {
+            long maxSequenceNumber = -1;
             tombstoneFile.open();
+
+            TombstoneFile rolloverFile = null;
+
             TombstoneFile.TombstoneFileIterator iterator = tombstoneFile.newIterator();
 
-            int count = 0, deleted = 0, copied = 0;
+            long count = 0, active = 0, copied = 0;
```

Comment:
qq, `deleted` -> `active` means we change the semantic?

Reply:
No, the old name meaning is wrong according to the context. The variable is used to track the number of tombstones which are still actively used, i.e. the deleted records are not cleaned up from data/index files by compaction job. ```copied``` and ```active``` are equal if ```isCleanUpTombstonesDuringOpen = true```, ```copied``` is 0 otherwise.